### PR TITLE
move the Translate, SubstrIndex, FindInSet functions to datafusion-functions

### LINKED ISSUE #9842: move the Translate, SubstrIndex, FindInSet functions to new datafusion-functions crate
### Is your feature request related to a problem or challenge?

As part of https://github.com/apache/arrow-datafusion/issues/9285 move the Translate, SubstrIndex, FindInSet functions to the new datafusion-functions crate in the new structure.

### Describe the solution you'd like

Functions are migrated to the new crate and use the new patterns as described in https://github.com/apache/arrow-datafusion/issues/9286 & https://github.com/apache/arrow-datafusion/pull/9216

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Omega359**
## Which issue does this PR close?

Closes #9842

## Rationale for this change

As part of https://github.com/apache/arrow-datafusion/issues/9285 the unicode functions should be migrated to the new datafusion-functions crate in the new structure

## What changes are included in this PR?

Code, tests

## Are these changes tested?

Yes

## Are there any user-facing changes?

No,

### CODE REVIEW

File: datafusion/core/Cargo.toml [hunk 1]
```diff
@@ -67,8 +67,6 @@ regex_expressions = [
 ]
 serde = ["arrow-schema/serde"]
 unicode_expressions = [
-    "datafusion-physical-expr/unicode_expressions",
```
Comment (alamb) [inline datafusion/core/Cargo.toml:70]: 🎉

File: datafusion/functions/src/unicode/substrindex.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        .zip(delimiter_array.iter())
+        .zip(count_array.iter())
+        .map(|((string, delimiter), n)| match (string, delimiter, n) {
+            (Some(string), Some(delimiter), Some(n)) => {
+                // In MySQL, these cases will return an empty string.
+                if n == 0 || string.is_empty() || delimiter.is_empty() {
+                    return Some(String::new());
+                }
+
+                let splitted: Box<dyn Iterator<Item = _>> = if n > 0 {
```
Comment (alamb) [inline datafusion/functions/src/unicode/substrindex.rs:115]: we can probably make this significantly faster by avoiding the `Box::new()` here and the call to `to_owned()` below. I realize this is not something changed in this PR

**Review by alamb — APPROVED** [review]
Thank you @Omega359  -- this looks great. I merged up from main to resolve a conflict

File: datafusion/functions/src/unicode/substrindex.rs (hunk 2 above)
Comment (Omega359) [reply datafusion/functions/src/unicode/substrindex.rs:115]: I'll create a separate benchmark for this.

**Comment by alamb** [comment]
Thanks again @Omega359

**Pull request merged**